File: cmasmas/uno/AddTwoNumbers/AddTwoNumbers.cpp
#include "AddTwoNumbers.h"
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class AddTwoNumbers
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode *result = new ListNode();
        ListNode *current = result;
        int carry = 0;

        while (l1 != nullptr || l2 != nullptr)
        {
            int sum = carry;
            if (l1 != nullptr)
            {
                sum += l1->val;
                l1 = l1->next;
            }
            if (l2 != nullptr)
            {
                sum += l2->val;
                l2 = l2->next;
            }
            carry = sum / 10;
            current->next = new ListNode(sum % 10);
            current = current->next;
        }
        if (carry > 0)
        {
            current->next = new ListNode(carry);
        }

        return result->next;
    }
    ListNode *addTwoNumbersRobada(ListNode *l1, ListNode *l2)
    {
        ListNode *dummyHead = new ListNode(0);
        ListNode *curr = dummyHead;
        int carry = 0;
        while (l1 != NULL || l2 != NULL || carry != 0)
        {
            int x = l1 ? l1->val : 0;
            int y = l2 ? l2->val : 0;
            int sum = carry + x + y;
            carry = sum / 10;
            curr->next = new ListNode(sum % 10);
            curr = curr->next;
            l1 = l1 ? l1->next : nullptr;
            l2 = l2 ? l2->next : nullptr;
        }
        return dummyHead->next;
    }
};

File: php/uno/resp/AddTwoNumbers.php
<?php
class ListNode {
    public $val = 0;
    public $next = null;
    function __construct(int $val = 0, ListNode $next = null) { 
        $this->val = $val; 
        $this->next = $next;
    }
}

class AddTwoNumbers {
    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function addTwoNumbers(ListNode $l1, ListNode $l2): ListNode {
        $carry = 0;
        $dummy = new ListNode(0);
        $curr = $dummy;
        while ($l1 != null || $l2 != null) {
            $x = ($l1 != null) ? $l1->val : 0;
            $y = ($l2 != null) ? $l2->val : 0;
            $sum = $x + $y + $carry;
            $carry = intval($sum / 10);
            $curr->next = new ListNode($sum % 10);
            $curr = $curr->next;
            if ($l1 != null) $l1 = $l1->next;
            if ($l2 != null) $l2 = $l2->next;
        }
        if ($carry > 0) {
            $curr->next = new ListNode($carry);
        }
        return $dummy->next;
    }
    function addTwoNumbers2(ListNode $l1, ListNode $l2): ListNode {
 
        $dummy = new ListNode();
        $current = $dummy;
        $carry = 0;

        while($l1 || $l2){

            $x = $l1 !== null ? $l1->val : 0;
            $y = $l2 !== null ? $l2->val : 0;
            
            $sum = $x + $y + $carry;

            $carry = (int) $sum / 10;
            $current->next = new ListNode($sum % 10);

            if($l1 !== null){
                $l1 = $l1->next;
            }

            if($l2 !== null){
                $l2 = $l2->next;
            }

            $current = $current->next;
        }

        if($carry >= 1){
            $current->next = new ListNode((int) $carry);
        }

        return $dummy->next;
        
    }
}